### ISSUE TITLE
Add option to have multiple custom gradients for one function

### ISSUE THREAD

**Issue by dpfau** [google/jax#3641]
At the moment, there are a few operations (determinants, LU decomposition) where we have to make a choice between a custom gradient that is simple and fast to compute, and one that is safe for all inputs - especially low-rank inputs. It may be opening a can of worms, but I think it may be worth it to extend custom gradients to include a parametric version of `defjvp` and `defvjp`, to allow something like this:

```
@custom_jvp
def my_func(x):
  ...

@my_func.defjvp(safe_det=False)
def _fast_jvp_my_func(primals, tangents):
  ...

@my_func.defjvp(safe_det=True)
def _safe_jvp_my_func(primals, tangents):
  ...

grad_func = jax.grad(my_func, safe_grad=True)
```

**Comment by dpfau** [comment]
I think this is something that only project members can do, but I assume this should be labeled "enhancement"

**Comment by hexahedria** [comment]
One question regarding this: does it make sense for `safe_grad` to be shared across the entire gradient transformation, as opposed to scoped to the particular primitive function you are calling? Perhaps there would be situations where you want to use safe grads for computing the determinant of A, but fast grads for computing the LU decomposition of B, and those happen to both happen in the same top-level function that you are computing grads for.

**Comment by shoyer** [comment]
I agree, this general idea sounds worthwhile. It's also a capability we want for ODE integration: https://github.com/google/jax/issues/2628

In practice, I think the right way to do this is probably to add JAX specific keyword arguments for controlling the implementing  how the gradient is calculated. For example: `jnp.linalg.det(x, grad_method='safe')` vs `jnp.linalg.det(x, grad_method='fast')`. These in turn might make use of entirely different JAX primitives.

**Comment by shoyer** [comment]
> One question regarding this: does it make sense for `safe_grad` to be shared across the entire gradient transformation, as opposed to scoped to the particular primitive function you are calling? Perhaps there would be situations where you want to use safe grads for computing the determinant of A, but fast grads for computing the LU decomposition of B, and those happen to both happen in the same top-level function that you are computing grads for.

For something like this, I would suggest adding multiple `grad_method` arguments at the top level in form, e.g., perhaps `det_grad_method='safe'` and `lu_grad_method='fast'`. But I'd really have to see an actual example to think it through. I suspect this would be relatively in JAX, though it might come up for some of the more complex SciPy functions.

**Comment by lockwo** [comment]
I see this is a pretty old issue, but has come up in some recent work of mine, has there been any updates on this idea?
